### PR TITLE
Export workspace to ADS on ALFView

### LINKED ISSUE #34823: ALFView - Toolbar button to save extracted workspace to ADS
**Description**
Add a button on the toolbar on the RHS of the ALFView interface which will save the workspace which is plotted in the PreviewPlot to the ADS. This will complete the refactor work which has decoupled the ALFView analysis pane from the ADS.

Next to the toolbar button there might be a line edit to specify the name to give the workspace.

### PR DESCRIPTION
**Pull request by robertapplin**
**Description of work.**
This PR adds the ability to export the plot to the ADS by placing a copy of the workspace in the ADS.

There is also the ability to open the plot in a separate window. This allows for a wider range of plotting options to be used, and the ability to save the plot to an image, etc.

**To test:**
Select ALF instrument in settings, and turn on data archive.
Open ALFView
Load run 82301
Click on Pick tab
Expand Rebin section. Rebin with 5.5,0.01,6
Click the Rectangle region selection tool at the top of the Pick tab
Select the two tubes in the centre with the yellow peak. The RHS plot should be updated.
Click Fit, and a Successful fit status label should appear.
Click the "Generate workspace from plot" button in the toolbar above the RHS plot. A workspace named "ALFView_exported" should appear in the ADS
Click the "Open plot in new window" button in the same toolbar. The plot should be opened in a new window. It is ok if the plot looks slightly different (e.g. the line colors etc. are different)

Fixes #34823

*This does not require release notes* because **release notes will be added in a later PR**

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by cailafinn — DISMISSED** [review]
Looks good and works well, just a couple of little things to mention:

File: qt/scientific_interfaces/Direct/ALFAnalysisView.cpp [hunk 1]
```diff
@@ -117,6 +117,14 @@ QWidget *ALFAnalysisView::createPlotWidget() {
...
 
 QWidget *ALFAnalysisView::createPlotToolbar() {
+  m_exportToADS = new QPushButton(MantidQt::Icons::getIcon("mdi.download"), "");
+  m_exportToADS->setToolTip("Generate workspace from plot. The workspace is named 'ALFView_exported'");
+  connect(m_exportToADS, SIGNAL(clicked()), this, SLOT(notifyExportWorkspaceToADSClicked()));
+
+  m_externalPlot = new QPushButton(MantidQt::Icons::getIcon("mdi.open-in-new"), "");
+  m_externalPlot->setToolTip("Open plot in new window. The new window has more plotting options.");
+  connect(m_externalPlot, SIGNAL(clicked()), this, SLOT(notifyExternalPlotClicked()));
+
```
[the remark this answers is not in the document]
Comment (robertapplin) [reply qt/scientific_interfaces/Direct/ALFAnalysisView.cpp:127]: If I remember correctly I think using the icons was an attempt to reduce the amount of information content on the GUI. Seeing as these buttons are accessories and not necessary for the main workflow, I thought it unimportant to give an explicit description unless the user specifically wants to know.

[the remark this answers is not in the document]
Comment (SilkeSchomann) [reply qt/scientific_interfaces/Direct/ALFAnalysisView.cpp:127]: Having buttons might confuse inexperienced users as they would normally not have to export the workspace or to display the plot in a new window so I think the icons help to keep the UI simple.

**Review by SilkeSchomann — DISMISSED** [review]
Happy to approve this PR. There were two change requests from a previous review. One was implemented but it was decided not to go ahead with the second one.

File: qt/scientific_interfaces/Direct/ALFAnalysisModel.cpp [hunk 2]
```diff
@@ -152,6 +160,34 @@ IFunction_sptr ALFAnalysisModel::calculateEstimate(MatrixWorkspace_sptr &workspa
...
+  // The ADS should not be used anywhere else apart from here. Note that a copy is exported.
+  if (auto const workspace = plottedWorkspace()) {
+    AnalysisDataService::Instance().addOrReplace(WS_EXPORT_NAME, workspace->clone());
+  }
+}
+
+void ALFAnalysisModel::openExternalPlot() const {
+  // Externally plot the extracted workspace or fitted workspace depending on which one is available.
+  if (auto const workspace = plottedWorkspace()) {
+    MantidQt::Widgets::MplCpp::plot({workspace}, boost::none, plottedWorkspaceIndices());
```
Comment (gemmaguest) [inline qt/scientific_interfaces/Direct/ALFAnalysisModel.cpp:173]: Is the model the right place to be doing the plotting? It feels like it breaks MVP a bit for the model to know about this other Qt widget (which is really a view concern). It also makes it hard to test. 

Would it be worth creating a little helper class like we do [here](https://github.com/mantidproject/mantid/blob/main/qt/scientific_interfaces/ISISReflectometry/GUI/Common/Plotter.h) in the reflectometry GUI? This means we can mock it and it also makes it easy to customise the plot.

Comment (robertapplin) [reply qt/scientific_interfaces/Direct/ALFAnalysisModel.cpp]: Good point, I've moved the external plotting code to the view. Has allowed me to add more tests so clearly this is the right thing to do 👍

File: qt/scientific_interfaces/Direct/test/ALFAnalysisPresenterTest.h [hunk 3]
```diff
@@ -205,8 +205,25 @@ class ALFAnalysisPresenterTest : public CxxTest::TestSuite {
...
-  void test_that_notifyExternalPlotClicked_calls_the_expected_model_function() {
-    EXPECT_CALL(*m_model, openExternalPlot()).Times(1);
+  void test_that_notifyExternalPlotClicked_will_open_an_external_plot_from_view() {
+    auto const workspace = WorkspaceCreationHelper::create2DWorkspace(1, 100);
+    std::vector<int> const workspaceIndices{0, 1};
+
+    EXPECT_CALL(*m_model, plottedWorkspace()).Times(1).WillOnce(Return(workspace));
+
+    EXPECT_CALL(*m_model, plottedWorkspaceIndices()).Times(1).WillOnce(Return(workspaceIndices));
+    EXPECT_CALL(*m_view, openExternalPlot(_, workspaceIndices)).Times(1);
```
Comment (gemmaguest) [inline qt/scientific_interfaces/Direct/test/ALFAnalysisPresenterTest.h:215]: Can we pass the workspace through here too? It's sometimes awkward dealing with the pointers but should be possible - let me know if you have problems. I often find that wrapping it in the `Eq` matcher explicity i.e. `Eq(workspace)` can resolve problems.

**Review by gemmaguest — DISMISSED** [review]
Thanks for the changes. Looks good and nice to have the extra tests :+1: Just one query about checking the workspace argument.

**Comment by robertapplin** [comment]
Looks like an unrelated test failure

**Comment by sf1919** [comment]
> Looks like an unrelated test failure

Yes it is. It seems PG3Analysis test can't run on our cloud linux nodes. They're temporarily offline to prevent this from happening again whilst we investigate the problem.

In the meantime it's probably not worth re-running this one if you might be pushing anymore commits. If you do push anymore commits it will run on an unaffected machine next time.

**Comment by gemmaguest** [comment]
Thanks, looks good :+1: @SilkeSchomann could you re-approve if you're happy with the latest changes then I'll merge it

**Review by SilkeSchomann — APPROVED** [review]
the latest changes look good so I am happy to approve again

**Pull request merged**